Refactor VDiff trait to make way for keyed list fixes

jstarry opened the pull request:
#### Description
Refactor `VDiff` trait and the diffing algorithm in preparation for keyed list fixes in https://github.com/yewstack/yew/pull/1231

#### Checklist:

- [x] I have ran `./ci/run_stable_checks.sh`
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works

<!-- Testing instructions -->
<!-- Check out the link below on how to setup and run tests -->
<!-- https://github.com/yewstack/yew/blob/master/CONTRIBUTING.md#test -->
<!-- If you're not sure how to test, let us know and we can provide guidance :) -->

<!-- Benchmark instructions -->
<!-- 1. Fork and clone https://github.com/yewstack/js-framework-benchmark -->
<!-- 2. Update `frameworks/yew/Cargo.toml` with your fork of Yew and the branch for this PR -->
<!-- 3. Open a new PR with your `Cargo.toml` changes -->
<!-- 4. Paste a link to the benchmark results: -->
<!-- - [ ] I have opened a PR against https://github.com/yewstack/js-framework-benchmark -->

jstarry commented:
@totorigolo do you mind looking over this? I pulled some of your changes out of your PR and made some modifications (removed the return type from `detach()`. I fixed the merge conflicts for your branch here: https://github.com/jstarry/yew/tree/vdiff-changes

totorigolo commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -340,162 +380,96 @@ impl VTag {
-            ancestor.captured.clear();
+        if let Some(ref mut ancestor_tag) = &mut ancestor_tag {
+            // Control the values of inputs and text area elements
+            ancestor_tag.refresh_value();
```
Shouldn't you move this after next line and change to this?
```suggestion
            self.refresh_value();
```

jstarry replied:
I'm not sure. I thought we needed to update the ancestors value to compare it against the value we want

totorigolo replied:
Indeed, you're right. Can you drop a comment to explain why it's needed?

jstarry replied:
Done

totorigolo commented on yew/src/virtual_dom/vtext.rs:
```diff
@@ -39,85 +37,44 @@ impl VText {
+                let text_node = self
+                    .reference
+                    .clone()
+                    .expect("text node should have reference");
```
```suggestion
                    .expect("Rendered VText nodes should have a ref");
```

jstarry commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -17,9 +17,6 @@ cfg_if! {
     /// The list of children nodes.
     pub children: Vec<VNode>,
-    /// Never use a placeholder element if set to true.
-    elide_placeholder: bool,
```
Good points! I removed this because I realized we could skip calling `apply` for `vtag.children` when it's empty

jstarry commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -101,13 +81,14 @@ impl VDiff for VList {
         }
+
         // Check for lefts to see if there are duplicates and show a warning.
         {
```
Sure! I'll add a comment. But just for the record I don't intend to ship the changes in this PR without your keyed list changes ;)

totorigolo replied:
Oh ok. But you could, this one doesn't really have to wait.

jstarry commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -123,55 +104,22 @@ impl VDiff for VList {
+
+        let first_left = lefts.next().expect("list should have at least one child");
+        let first_child = first_left.apply(parent_scope, parent, next_sibling, rights.next());
+        next_sibling = Some(first_child.clone());
```
Ah you're so right. Might need to iterate backwards then.

totorigolo replied:
I see two solutions:

1. `next_sibling = node.next_sibling()`
2. ... wait, I _just_ realized: we can repetitively insert before `next_sibling`, no need to mutate it. The next sibling is a constant inside of the reconciliation loop (at least for the simple non-keyed version).

jstarry replied:
It's not quite a constant when updating an existing list. I went with the reverse iteration approach

jstarry commented:
Thanks for the thorough review! I will try to fix this up tomorrow, great suggestions :)

jstarry commented:
> Also, input/textarea virtual tag nodes deserve a test, since that have a special treatment.

There are tests for that already fyi: `check_input_current_value_sync`

I added one of your non-keyed tests @totorigolo and expanded on it

jstarry commented on yew/src/virtual_dom/vnode.rs:
```diff
@@ -31,6 +32,35 @@ pub enum VNode {
     VRef(Node),
 }
 
+impl From<&VNode> for Node {
```
@totorigolo this is a slight modification of your `reference` method. I'm not sure I love the ergonomics of using the `From` trait so I'm open to changing. I needed this because we don't have `next_sibling` when mounting a component. We need to get it from the ancestor.

jstarry commented on yew/src/virtual_dom/mod.rs:
```diff
@@ -176,60 +176,78 @@ enum Patch<ID, T> {
-    /// The reference _may still be mutated_.
-    Keep,
+/// Delay detachment of ancestors until after inserting new nodes
+struct DelayDetach<'a> {
```
This allows us to insert a new node before the ancestor before detaching it. This way, we don't need `detach()` to return the `next_sibling` (which saves us a DOM API call)

totorigolo commented on yew/src/virtual_dom/vtext.rs:
```diff
@@ -52,24 +52,27 @@ impl VDiff for VText {
 
-            ancestor.detach(parent);
-        }
+            next_sibling = Some((&ancestor).into());
+            Some(super::DelayDetach { ancestor, parent })
```
In some way, `next_sibling` is a pointer, and it becomes invalid as soon as `DelayedDetach` is dropped. We could define its lifetime as being valid until `_ancestor` is dropped. Our course, using it afterwards doesn't result in a segfault, but in correctness issue. So I wonder if there is a way to enforce it or advertise it a bit more. I am mainly worried about later refactoring that could go wrong.

Could we instead give the delayed dropping responsibility to next_sibling? Like creating an enum for next_sibling whose Drop implementation implements DelayedDetach for one of its variants? Or maybe that's too overkill?

jstarry replied:
That's a good idea, I agree that this implementation is quite fragile. I'll play around with your ideas and see if I can get something working well

jstarry replied:
I overhauled things so that we don't need to do this delayed dropping anymore

totorigolo commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -94,77 +94,211 @@ impl VDiff for VList {
-        let first_left = lefts.next().expect("list should have at least one child");
-        let first_child = first_left.apply(parent_scope, parent, next_sibling, rights.next());
-        next_sibling = Some(first_child.clone());
+        // Detach rights until length equals lefts.len()
```
```suggestion
        // Detach rights, in reverse order, until length equals lefts.len()
```

totorigolo commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -94,77 +94,211 @@ impl VDiff for VList {
 
-        for left in lefts {
-            next_sibling = Some(left.apply(parent_scope, parent, next_sibling, rights.next()));
+        // Apply lefts until length equals rights.len()
```
```suggestion
        // Apply lefts, in reverse order, until length equals rights.len()
```

totorigolo commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -94,77 +94,211 @@ impl VDiff for VList {
 
-        for mut right in rights {
-            right.detach(parent);
+        // Diff children in reverse order to ensure next_sibling is updated properly
```
I had the same idea, and I think that this is a good idea, but I wonder if the fact that components will be created/changed in reverse order can be problematic. I think not since there is no guarantee anywhere, but at least IMHO this needs to be marked as potential breaking change in the release note.

jstarry replied:
Good point, I didn't consider the component mounting order

jstarry replied:
I ended up scrapping the reverse order approach

totorigolo commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -94,77 +94,211 @@ impl VDiff for VList {
+        let parent_scope: AnyScope = Scope::<Comp>::new(None).into();
+        let parent_element = crate::utils::document().create_element("div").unwrap();
+
+        let scheduler_lock = scheduler.lock();
```
Could you add a comment to explain why this lock is needed?

jstarry replied:
Wasn't needed, removed!

jstarry commented on yew/src/virtual_dom/vnode.rs:
```diff
@@ -31,6 +32,35 @@ pub enum VNode {
+                .node_ref
+                .get()
+                .expect("VComp should always wrap a node"),
+            VNode::VList(vlist) => (&vlist.children[0]).into(),
```
Good call, I'll change this to a method with naming that makes it clearer

jstarry replied:
Changed to `first_node()`

jstarry commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -431,18 +431,23 @@ impl VDiff for VTag {
             // Preserve the reference that already exists.
             self.reference = ancestor_tag.reference.take();
-        } else {
+        } else if self.reference.is_none() {
```
Yeah, if ancestor is None

jstarry commented:
@totorigolo I ran some benchmarks on this change and they look pretty good!
https://github.com/yewstack/js-framework-benchmark/pull/4#issuecomment-640088862

It's been great chipping away at this with you! I'll get this PR wrapped up this weekend so that your keyed changes are fully unblocked :)

jstarry commented:
Still got a ways to go though until matching `wasm-bindgen` performance!

totorigolo commented:
> Still got a ways to go though until matching `wasm-bindgen` performance!

wasm-bindgen doesn't have a virtual DOM, so it only serves to see the overhead of calling DOM functions. Said another way, yew minus wasm-bindgen represents the overhead of React's virtual DOM/reconciliation model.

jstarry commented:
> Said another way, yew minus wasm-bindgen represents the overhead of React's virtual DOM/reconciliation model.

Exactly! Hope we can minimize it

jstarry commented:
@totorigolo do you want to take another look at this? I believe I addressed all of your comments. I also overhauled the code more and added lot of tests!

jstarry commented:
<img width="135" alt="Screenshot 2020-06-09 00 12 17" src="https://user-images.githubusercontent.com/1076145/84054425-f61e3080-a9e5-11ea-8132-f322f1e367f2.png">

Didn't move the needle too much after all 🤷

totorigolo commented on yew/src/virtual_dom/vlist.rs:
```diff
@@ -94,67 +94,67 @@ impl VDiff for VList {
 
-        first_child
+        assert!(!nodes.is_empty(), "VList should have at least one child");
+        nodes.remove(0)
```
Use `swap_remove`.

jstarry replied:
Oh great call, done!

jstarry commented on yew/src/virtual_dom/vnode.rs:
```diff
@@ -32,6 +33,34 @@ pub enum VNode {
+                .node_ref
+                .get()
+                .expect("VComp should always wrap a node"),
+            VNode::VList(vlist) => (&vlist.children[0]).first_node(),
```
Didn't intend to make it public, I've changed to `pub(crate)` visibility. Thanks!
